File: Data Structures & Algorithms/duplicate-integer/submission-1.cpp
class Solution {
public:
    bool hasDuplicate(vector<int>& nums) {
        std::unordered_map<int, int> m;
        for (int num : nums) {
            if (m[num] >= 1) return true;
            m[num]++;
        }
        return false;
    }
};

/*
*/

File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        res = []
        def dfs (s, arr):
            #print(f"dfs({s},{arr})")
            if len(arr) == len(nums):
                #print(f"we append {arr} to {res}")
                res.append(arr)
                #print(res)
                return
            for num in nums:
                if num not in s:
                    #print(f"{num} not in {s}")
                    s.add(num)
                    arr.append(num)
                    dfs(s, arr.copy())
                    s.remove(num)
                    arr.remove(num)
        s = set()
        dfs(s, [])
        return res
